### PR TITLE
Resolve completion of module options and sub options based on their type

### PR DESCRIPTION
**Pull request by priyamsahoo**
This PR adds the feature to resolve the module options and sub options based on their types by providing the following suffixes so that it is easier for the user to guess the type of value and and at same time, uninterrupted auto-completions for subsequent sub options are provided.

| **Type** | **Suffix** | **Explanation** |
| :---: | :---: | :---: |
| List | :\r\t-  | move to next line and set cursor and providing a tab and "-" | 
| Dictionary | :\r\t\t | move to next line and set cursor after double tab |
| All other cases | : | set cursor after a space |

Appropriate tests are added under `test/providers/completionResolver.test.ts`, in continuation of #273.

### CODE REVIEW

File: src/providers/completionProvider.ts
Comment (ssbarnea) [inline src/providers/completionProvider.ts:348]: This might brake in the future based on editor configuration, as it might be configured to auto-indent. still, no idea on how to do this better, just an observation.

**Review by ganeshrn — APPROVED** [review]

**Comment by priyamsahoo** [comment]
@yaegassy, could you try this change with you extension and let us know how this works ?

File: src/providers/completionProvider.ts
Comment (tomaciazek) [inline src/providers/completionProvider.ts:399]: Same here - I'm ending up with extra indentation.

**Review by tomaciazek — CHANGES_REQUESTED** [review]
Nice idea, though I've discovered some issues. Also, is this what users really want? What if they want to pass variables? Then this automatic line jump is going to drive them mad :D

Though in case of resolving module names, it does make a lot of sense.

**Comment by yaegassy** [comment]
@priyamsahoo I tried this with an extension for my Vim. Unfortunately, the `\r` did not break the line...

**DEMO(mp4)

https://user-images.githubusercontent.com/188642/160057638-64e1d90e-13cc-4ed8-acd7-0713a9f54f65.mp4

**Comment by yaegassy** [comment]
Just in case you are interested, here is a link to yaml-langauge-server's yamlCompletion.ts. It might be of some help. <https://github.com/redhat-developer/yaml-language-server/blob/main/src/languageservice/services/yamlCompletion.ts>

**Comment by priyamsahoo** [comment]
Thanks for the update @yaegassy. Will let you know once I push commits to fix this issue.

File: src/providers/completionProvider.ts
Comment (priyamsahoo) [reply src/providers/completionProvider.ts:382]: Thanks for pointing it out. A default value has been added now.

File: src/providers/completionProvider.ts
Comment (priyamsahoo) [reply src/providers/completionProvider.ts]: So the idea behind the extra `\t` is as follows:

If a dict is a first element of a list then only `\r\t` would put the cursor at the next line under the same level of the option which is not correct (since, we want to place the cursor one level deep of that option). This is because of the extra `-` that comes from the list.

But if the option is not the first element of the list, it works fine.

So, to overcome this problem, dicts are suffixed with `\r\t\t` (double tabs).

Anyways, I have added logic to identify if an option is the first element of the list or not. And accordingly indentations are provided. You can now take a look.

**Comment by priyamsahoo** [comment]
@tomaciazek, @yaegassy take a look at the changes now and let me know if they work correctly.

**Comment by priyamsahoo** [comment]
> Nice idea, though I've discovered some issues. Also, is this what users really want? What if they want to pass variables? Then this automatic line jump is going to drive them mad :D
> 
> Though in case of resolving module names, it does make a lot of sense.

I added this feature because people asked for it :), especially people dealing with network modules. This feature enhances the playbook writing experience at lot when there are nested module options.

Plus, I will wait for people to complain for something that annoys them. Feature enhancement is a continuous process :)

File: src/providers/completionProvider.ts
Comment (tomaciazek) [reply src/providers/completionProvider.ts]: Works like a charm now :)

**Comment by tomaciazek** [comment]
Could I then ask you to implement a feature toggle for this? If you do, I think that only options/sub-options need that - automatic newline after module should not bother anyone.

I understand that feature enhancement is a process, but some things are just foreseeable, and we can prevent some bad reviews if we allow people to choose ;)

**Comment by ssbarnea** [comment]
> Could I then ask you to implement a feature toggle for this? If you do, I think that only options/sub-options need that - automatic newline after module should not bother anyone.
> 
> I understand that feature enhancement is a process, but some things are just foreseeable, and we can prevent some bad reviews if we allow people to choose ;)

@tomaciazek We need to be very careful about how much customization we allow, mainly because it would be extremely easy to end-up not being able to cope with its complexity. I would personally support only customizations that are essential, or which do not make the logic or testing much harder.

To give an idea: many options do explode testing matrix complexity as you will have to test on they interact with others.

**Review by tomaciazek — APPROVED** [review]
That's what unit tests are for - explosion of variants happens on a level above them: integration / UI tests. Still, I won't block this PR anymore :)

**Comment by yaegassy** [comment]
@priyamsahoo One point I would like to ask you to check. Is this feature working correctly in `vscode-ansible`? If possible, a screen capture (mp4) would be appreciated.

**Comment by priyamsahoo** [comment]
Sure @yaegassy.

https://user-images.githubusercontent.com/42550351/161212996-c7832adb-2fc8-4cb0-9ddf-2ad3e56177d2.mp4

You can see the indentation and the syntax completion it provides.

Is there any issue that you are facing ?

**Comment by yaegassy** [comment]
Sorry to confirm this after the PR merge, but unfortunately the indentation is not correct in Vim/Neovim.

I tried with "coc-ansible", which is my maintenance system. I also tried with "Neovim built-in LSP" and the situation is the same in both cases.

**DEMO(mp4)**:

https://user-images.githubusercontent.com/188642/161216267-7c243248-a1e3-44b2-b6fc-ac0bc0bc6da1.mp4

**Comment by ssbarnea** [comment]
@yaegassy Can you come up with a fix? We can make as many releases as needed as that is a library and each consumer application is expected to test new versions before starting to consume them.

**Comment by priyamsahoo** [comment]
> @priyamsahoo Sorry to confirm this after the PR merge, but unfortunately the indentation is not correct in Vim/Neovim.
> 
> I tried with "coc-ansible", which is my maintenance system. I also tried with "Neovim built-in LSP" and the situation is the same in both cases. :(
> 
> **DEMO(mp4)**:
> 
>  coc-ansible-als-0-5-4.mp4

@yaegassy, I want you to try one last thing on your system. In this PR, can you locally change `${EOL}` to `\n` wherever they are mentioned in the `completionProvider.ts`.
I believe this should work.

**Comment by priyamsahoo** [comment]
@yaegassy, the thing is, from the log I can confirm that the feature from the server side is working correctly. It is sending `\n\t`, but the problem is how the extension is handling it. The extension is doing `\n` and then `\t` but while doing a `\n` (new line), the cursor is sent to the column 0 of the new line. Rather, it should move to the next line on same indentation (and thus `\t` should have made it one level deeper).

It would be good if you could find the appropriate escape character to move to new line at same indentation for vim/nvim(which automatically it should do).

**Comment by yaegassy** [comment]
@priyamsahoo These are completion items for `yaml-language-server`. A "space" is used instead of `\t`. The indentation works correctly in yaml-language-server, so there may be something that can be adjusted in ALS.

```
        {
            "kind": 10,
            "label": "ports",
            "insertText": "ports:\n  - ",
            "insertTextFormat": 2,
            "documentation": "",
            "textEdit": {
                "range": {
                    "start": {
                        "line": 5,
                        "character": 4
                    },
                    "end": {
                        "line": 5,
                        "character": 4
                    }
                },
                "newText": "ports:\n  - "
            }
        },
```

https://user-images.githubusercontent.com/188642/161236640-152e757f-613b-4467-b877-1e2d44e92bd4.mp4

**Comment by priyamsahoo** [comment]
@yaegassy, I am looking into it. But the problem is it is working properly in vscode and I am unable to test these on vim extn.
It would be good, if you could try implementations. (Maybe, spaces in place of tabs, etc. and see what works well with vim/nvim extension).

I would be happy to discuss on this and review PR if you come up with one.

**Comment by yaegassy** [comment]
@priyamsahoo OK, I found the cause and will do a PR.

**fixed**:

https://user-images.githubusercontent.com/188642/161245925-6a16c84c-f2f2-4796-b4cc-4925ad95ed80.mp4

**Comment by priyamsahoo** [comment]
> @priyamsahoo OK, I found the cause and will do a PR.
> 
> **fixed**:
> 
>  Kapture.2022-04-01.at.19.25.34.mp4

Awesome !

**Comment by yaegassy** [comment]
PRed https://github.com/ansible/ansible-language-server/pull/285

**Pull request merged**
